core/GlobalState.cc

```diff
@@ -926,7 +938,30 @@ SymbolRef GlobalState::enterTypeMember(Loc loc, SymbolRef owner, NameRef name, V
     }
 
     flags = flags | Symbol::Flags::TYPE_MEMBER;
-    SymbolRef result = enterSymbol(loc, owner, name, flags);
+
+    SymbolData ownerScope = owner.dataAllowingNone(*this);
+    histogramInc("symbol_enter_by_name", ownerScope->members().size());
+
+    auto &store = ownerScope->members()[name];
+    if (store.exists()) {
+        ENFORCE((store.data(*this)->flags & flags) == flags, "existing symbol has wrong flags");
+        counterInc("symbols.hit");
+        return store;
+    }
+
+    ENFORCE(!symbolTableFrozen);
+    auto result = SymbolRef(this, SymbolRef::Kind::TypeMember, typeMembers.size());
+    store = result; // DO NOT MOVE this assignment down. emplace_back on typeMembers invalidates `store`
+    typeMembers.emplace_back();
+
+    SymbolData data = result.dataAllowingNone(*this);
+    data->name = name;
+    data->flags = flags;
+    data->owner = owner;
+    data->addLoc(*this, loc);
+    DEBUG_ONLY(categoryCounterInc("symbols", "type_member"));
+    wasModified_ = true;
+
```

Comment:
Yeah, these are inlined. Dmitry requested it above, in one of the first reviews.

I figured that this will be beneficial when we actually have per-symbol-type structs, which will have separate `enter` logic anyway. This is only inlined once per symbol type, right?